File: 0873-length-of-longest-fibonacci-subsequence/0873-length-of-longest-fibonacci-subsequence.js
/**
 * @param {number[]} arr
 * @return {number}
 */
var lenLongestFibSubseq = function(arr) {
    const dp = {};
    for (let i = 0; i < arr.length - 1; i++) {
        dp[arr[i]] = dp[arr[i]] || {};
        for (let j = i + 1; j < arr.length; j++) {
            dp[arr[i]][arr[j]] = 2;
        }
    }
    for (let j = 0; j < arr.length; j++) {
        for (let k = j + 1; k < arr.length; k++) {
            const diff = arr[k] - arr[j];
            if (diff < arr[j] && dp[diff]) dp[arr[j]][arr[k]] = dp[diff][arr[j]] + 1;
        }
    }
    let ans = 0;
    for (const outerKey in dp) {
        for (const innerKey in dp[outerKey]) {
            if (dp[outerKey][innerKey] > 2) ans = Math.max(ans, dp[outerKey][innerKey]);
        }
    }
    return ans;
};

File: 0873-length-of-longest-fibonacci-subsequence/0873-length-of-longest-fibonacci-subsequence.py
from collections import defaultdict

class Solution:
    def lenLongestFibSubseq(self, arr: List[int]) -> int:
        dp = defaultdict(dict)
        for i in range(len(arr)-1):
            for j in range(i+1, len(arr)):
                dp[arr[i]][arr[j]] = 2
        for j in range(len(arr)):
            for k in range(j+1, len(arr)):
                diff = arr[k]-arr[j]
                if diff < arr[j] and diff in dp:
                    dp[arr[j]][arr[k]] = dp[diff][arr[j]]+1
        ans = 0
        for outerKey in dp:
            for innerKey in dp[outerKey]:
                innerVal = dp[outerKey][innerKey]
                if innerVal > 2:
                    ans = max(ans, innerVal)
        return ans